### ISSUE TITLE
Vet SciPy for use of `long` as the default integer

### ISSUE THREAD

**Issue by seberg** [scipy/scipy#19462]
We just merged https://github.com/numpy/numpy/pull/24224 changing the default integer in NumPy.  This changes the definition of `np.int_` to intp, which only matters on windows, though.

There are three things that might be good to ensure here:
1. In some cases `np.int_` is probably used as long, and needs to be `np.dtype("long").type` (compatible version.
2. I think I saw a cython use `np.int_t`, which is ill defined unfortunately. It needs to be either `npy_long` (because it is hard to create a compatible name).  Or `intp` (with potential casting if required).  Worst case, you need to support both, which would be annoying I admit (fused type).
3. It probably doesn't matter much, but I think many special functions are defined with long loops.  In principle it may make sense to make them int or int64 loops to ensure 64bit integers can match them.

It is very likely that this will break something on windows at this moment, the Cython issue should break compilation.

**Comment by tylerjereddy** [comment]
Some of this is dealt with in gh-19466, now merged, but there are a few broader points above so leaving this open for now.

**Comment by tylerjereddy** [comment]
I think we fixed most of the issues here for `1.13.0`, in large part by kicking the can down the road with `np.dtype("long")`.

That said, there is a point about `special` function loops types above that I'm not tracking, and we have lots of C++ rewrite activity happening in `special` now, so I'll bump the milestone in case that particular point remains relevant.

**Comment by ev-br** [comment]
Would be great to act on the remaining `np.dtype("long")` uses:

```
$ git grep 'np.dtype("long")'
scipy/interpolate/_rbfinterp.py:    out = np.zeros((nmonos, ndim), dtype=np.dtype("long"))
scipy/linalg/_decomp_schur.py:        a1 = asarray(a, dtype=np.dtype("long"))
scipy/optimize/_trlib/_trlib.pyx:        self.iwork = np.zeros([iwork_size], dtype=np.dtype("long"))
scipy/optimize/_trlib/_trlib.pyx:                               dtype=np.dtype("long"))
scipy/special/_basic.py:            dt = np.dtype("long")
scipy/special/_lambertw.py:    k = np.asarray(k, dtype=np.dtype("long"))
scipy/special/_spherical_bessel.py:    n = np.asarray(n, dtype=np.dtype("long"))
scipy/special/_spherical_bessel.py:    n = np.asarray(n, dtype=np.dtype("long"))
scipy/special/_spherical_bessel.py:    n = np.asarray(n, dtype=np.dtype("long"))
scipy/special/_spherical_bessel.py:    n = np.asarray(n, dtype=np.dtype("long"))
scipy/special/tests/test_orthogonal.py:    n = np.arange(N, dtype=np.dtype("long"))
scipy/special/tests/test_orthogonal_eval.py:    n = np.arange(0, 10000, 7, dtype=np.dtype("long"))
scipy/special/tests/test_orthogonal_eval.py:            p = (p[0].astype(np.dtype("long")),) + p[1:]
scipy/special/tests/test_sf_error.py:    args = np.asarray(args, dtype=np.dtype("long"))
```

Meanwhile, this is definitely not release-critical, thus bumping the milestone.

**Comment by lucascolley** [comment]
x-ref https://github.com/scipy/scipy/pull/19724#discussion_r1434315899

**Comment by lucascolley** [comment]
@seberg does https://github.com/scipy/scipy/blob/d5d3faa73f9822859851bdb0a6f5c936b38a5b2f/scipy/sparse/csgraph/parameters.pxi#L6 fall under the scope of this issue? That seems related to gh-24629

**Comment by seberg** [comment]
No, this was very narrow on "long" specific loops that could notice a change because NumPy 2 changed windows default integers to "long long" (64 bit).

I.e. ufuncs that have a loop like `fl->f` but not `fq->f`.

In practice that will be fixed now with explicit `n = np.asarray(n, dtype=np.dtype("long"))` that Evgeni mentioned.

So I think we can either close the issue, or consider it an issue about cleaning up explicit long casts.

EDIT: OK, also Cython, but this one is explicitl `int32_t`, so there was no real change w.r.t. to this issue.

**Comment by ev-br** [comment]
> In practice that will be fixed now with explicit n = np.asarray(n, dtype=np.dtype("long")) that Evgeni mentioned.

 `dtype=np.dtype("long")` was a temporary crutch to paper over the numpy 1/2 split. Now that scipy only support numpy 2.x, I'd really hope we get rid of these instead.
